substrate/frame/revive/src/lib.rs

```diff
@@ -130,20 +130,47 @@ const SENTINEL: u32 = u32::MAX;
 const LOG_TARGET: &str = "runtime::revive";
 
 pub(crate) mod eth_block_storage {
+	use crate::{evm::block_hash::AccumulateReceipt, EventLog, H160, H256};
+	use alloc::vec::Vec;
+	use alloy_core::primitives::Bloom;
 	use environmental::environmental;
 
 	/// The maximum number of block hashes to keep in the history.
 	pub const BLOCK_HASH_COUNT: u32 = 256;
 
-	// Indicates whether an Ethereum call is currently being executed.
-	environmental!(executing_call: bool);
+	// The events emitted by this pallet while executing the current inflight transaction.
+	//
+	// The events are needed to reconstruct the receipt root hash, as they represent the
+	// logs emitted by the contract. The events are consumed when the transaction is
+	// completed. To minimize the amount of used memory, the events are RLP encoded directly.
+	environmental!(receipt: AccumulateReceipt);
 
-	pub fn is_executing_ethereum_call() -> bool {
-		executing_call::with(|ctx| *ctx).unwrap_or(false)
+	/// Capture the Ethereum log for the current transaction.
+	///
+	/// This method does nothing if called from outside of the ethereum context.
+	pub fn capture_ethereum_log(contract: H160, data: &[u8], topics: &[H256]) {
+		receipt::with(|receipt| {
+			receipt.add_log(EventLog { contract, data: data.to_vec(), topics: topics.to_vec() });
```

Comment:
It seems unnecessary to clone the data here if you re-encode it anyways. Should be possible from a reference.

Reply:
Yep makes sense! Have added a new custom RLP encoding for the logs as well 🙏